src/hp/magicsock/derp_actor.rs

```diff
@@ -0,0 +1,810 @@
+    /// home connection, or what was once our home), then we remember that route here to optimistically
+    /// use instead of creating a new DERP connection back to their home.
+    derp_route: HashMap<key::node::PublicKey, DerpRoute>,
+    msg_sender: flume::Sender<ActorMessage>,
+}
+
+impl DerpActor {
+    pub(super) fn new(
+        conn: Arc<Inner>,
+        receiver: mpsc::Receiver<DerpActorMessage>,
+        msg_sender: flume::Sender<ActorMessage>,
+    ) -> Self {
+        DerpActor {
+            conn,
+            receiver,
+            active_derp: HashMap::default(),
+            derp_route: HashMap::new(),
+            msg_sender,
+        }
+    }
+
+    pub(super) async fn run(mut self) {
+        let mut cleanup_timer = time::interval_at(
+            time::Instant::now() + DERP_CLEAN_STALE_INTERVAL,
+            DERP_CLEAN_STALE_INTERVAL,
+        );
+
+        loop {
+            tokio::select! {
+                Some(msg) = self.receiver.recv() => {
+                    match msg {
+                        DerpActorMessage::Send { region_id, contents, peer } => {
+                            self.send_derp(region_id, contents, peer).await;
+                        }
+                        DerpActorMessage::Connect { region_id, peer } => {
+                            self.connect_derp(region_id, peer.as_ref()).await;
+                        }
+                        DerpActorMessage::CloseOrReconnect { region_id, reason } => {
+                            self.close_or_reconnect_derp(region_id, reason).await;
+                        }
+                        DerpActorMessage::AddPeerRoutes { peers, region, derp_client } => {
+                            self.add_derp_peer_routes(peers, region, derp_client);
+                        }
+                        DerpActorMessage::RemovePeerRoutes { peers, region, derp_client } => {
+                            self.remove_derp_peer_routes(peers, region, &derp_client);
+                        }
+                        DerpActorMessage::NotePreferred(my_derp) => {
+                            self.note_preferred(my_derp).await;
+                        }
+                        DerpActorMessage::SetDerpMap(derp_map) => {
+                            self.set_derp_map(derp_map).await;
+                        }
+                        DerpActorMessage::MaybeCloseDerpsOnRebind(ifs) => {
+                            self.maybe_close_derps_on_rebind(&ifs).await;
+                        }
+                        DerpActorMessage::Shutdown => {
+                            debug!("shutting down");
+                            self.close_all_derp("conn-close").await;
+                            break;
+                        }
+                    }
+                }
+                _ = cleanup_timer.tick() => {
+                    trace!("tick: cleanup");
+                    self.clean_stale_derp().await;
+                }
+                else => {
```

Comment:
Nice. I didn't know that you can pattern match in `select!`. So this is for when self.receiver.recv() returns None?

Reply:
(going off to change some code in my interprocess stuff now...)

Reply:
> So this is for when self.receiver.recv() returns None?

this will catch anything not matched by previous statements, so in this case `None`